rclpy/src/rclpy/_rclpy_qos_event.c

```diff
@@ -0,0 +1,363 @@
+  * \param[in] pyparent Capsule containing the parent Publisher or Subscription.
+  * \return capsule containing rcl_event_t.
+  * \return NULL on failure.
+  */
+static PyObject *
+rclpy_create_event(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  unsigned PY_LONG_LONG pyevent_type;
+  PyObject * pyparent = NULL;
+
+  rcl_ret_t ret;
+  rcl_subscription_t * subscription = NULL;
+  rcl_publisher_t * publisher = NULL;
+  rcl_event_t * event = NULL;
+
+  PyObject * pyevent = NULL;
+
+  if (!PyArg_ParseTuple(args, "KO", &pyevent_type, &pyparent)) {
+    return NULL;
+  }
+
+  if (_is_pycapsule_rcl_subscription(pyparent)) {
+    subscription = _pycapsule_to_rcl_subscription(pyparent);
+  } else if (_is_pycapsule_rcl_publisher(pyparent)) {
+    publisher = _pycapsule_to_rcl_publisher(pyparent);
+  } else {
+    PyErr_Format(PyExc_TypeError, "Event parent was not a valid Publisher or Subscription.");
+    return NULL;
+  }
+
+  event = _new_zero_initialized_rcl_event();
+  if (!event) {
+    return NULL;
+  }
+
+  if (subscription) {
+    ret = rcl_subscription_event_init(event, subscription, pyevent_type);
+  } else {
+    ret = rcl_publisher_event_init(event, publisher, pyevent_type);
+  }
+  if (!_check_rcl_return(ret, "Failed to initialize event")) {
+    PyMem_Free(event);
+    return NULL;
+  }
+
+  pyevent = PyCapsule_New(event, "rcl_event_t", _destroy_event_capsule);
+  if (!pyevent) {
+    ret = rcl_event_fini(event);
+    PyMem_Free(event);
+    _check_rcl_return(ret, "Failed to fini 'rcl_event_t'");
+    return NULL;
+  }
+
+  return pyevent;
+}
+
+/// Get a pending QoS event's data.
+/**
+  * After having determined that a middleware event is ready, get the callback payload.
+  *
+  * Raises RuntimeError on failure to take the event from the middleware.
+  * Raises TypeError if the capsules are not the correct types.
+  * Raises ValueError on unknown event_type argument.
+  *
+  * \param[in] pyevent Event handle from rclpy_create_event.
+  * \param[in] pyevent_type Enum value of
+  *   rcl_publisher_event_type_t or rcl_subscription_event_type_t, chosen by the type of pyparent.
+  * \param[in] pyparent Capsule containing the parent Publisher or Subscription.
+  * \return Python object from rclpy.qos_event containing callback data.
+  * \return NULL on failure.
+  */
+static PyObject *
+rclpy_take_event(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  // Arguments
+  PyObject * pyevent = NULL;
+  PyObject * pyparent = NULL;
+  unsigned PY_LONG_LONG pyevent_type;
+
+  // Type conversion
+  rcl_ret_t ret;
+  rcl_event_t * event = NULL;
+  PyObject * pyqos_event = NULL;
+  _qos_event_callback_data_t event_data;
+  _qos_event_data_filler_function event_filler = NULL;
+
+  if (!PyArg_ParseTuple(args, "OOK", &pyevent, &pyparent, &pyevent_type)) {
+    return NULL;
+  }
+
+  if (!_is_pycapsule_rcl_event(pyevent)) {
+    PyErr_Format(PyExc_TypeError, "Capsule was not a valid rcl_event_t");
```

Comment:
nit, would rather check if `_pycapsule_to_rcl_event` returned NULL instead of calling `PyCapsule_IsValid` since it's shorter and the capsule name only has to be given in one place. Though one thing to note if you're catking this exception is `PyCapsule_GetPointer` raises `ValueError` instead of `TypeError`.